### PR TITLE
Create a function to append different query params to internal and external URLs

### LINKED ISSUE #8516: Create a function to append different query params to internal and external URLs
This supports #8249 

For #8249, we need a way to differentiate between internal and external URLs, and to append a different set of query params to each. It seems like a function that accepts a URL string, an `internalQueryParams` object and an `externalQueryParams` object, and then returns the appropriate URL would be a good helper for this use case, so that's what this issue is about.

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #8516 

The tests seem pretty repetitive, but I think it's important to be explicit about the data in each test so they can easily be reviewed and verified that they are testing the correct thing, so I'm not sure I can cut down on the repetitiveness with an extra helper.

### CODE REVIEW

**Comment by bobsilverberg** [comment]
> Can you give me some examples of how you envision using this function? I'm just not sure that we need it.
> 

@kumar303 I guess at this point we only need it for the URLs for `cta`s  coming from the `hero` API, but I imagined we might need it elsewhere as well. There is a requirement for the Secondary hero shelf that we append a certain set of params to internal links and a different set of params to external links. The reason I thought to make it a single function which both checks for an internal vs. external URL and also appends different URL params in each case is because we need access to the same data for both purposes (which we get via `url.parse`).

I could just do the check for internal vs. external in the Secondary shelf component, but then I'd be calling `url.parse` in that function, and it would again be called in the `addQueryParams()` function. I know this is likely not a performance concern, but it just seemed better to only be doing it once.

I also broke this out into a function in order to have a nice small PR for just this feature/logic, as opposed to including it in the larger PR that I'm working on for the Secondary shelf.

I still feel like this is a decent feature to have as a standalone utility function, but if you feel strongly that we shouldn't have this then I'll abandon this and move it into the code for the Secondary hero shelf. We're still probably going to need all of the tests though.

File: tests/unit/amo/utils/test_index.js [hunk 1]
```diff
@@ -13,4 +15,211 @@ describe(__filename, () => {
...
+
+      expect(
+        _makeInternalExternalURL({
+          internalQueryParams,
+          urlString,
+        }),
+      ).toEqual(expectedURL);
+    });
+
+    it('adds internal query params to an absolute URL for the host', () => {
```
Comment (kumar303) [inline tests/unit/amo/utils/test_index.js]: I think I know what this means but it tripped me up and I had to read it several times. I suggest trying to reword it.

File: tests/unit/amo/utils/test_index.js [hunk 2]
```diff
@@ -13,4 +15,211 @@ describe(__filename, () => {
...
+
+      expect(
+        _makeInternalExternalURL({
+          internalQueryParams,
+          urlString,
+        }),
+      ).toEqual(expectedURL);
+    });
+
+    it('overrides internal query params for a relative URL', () => {
```
Comment (kumar303) [inline tests/unit/amo/utils/test_index.js]: If you switch the implementation to [`addQueryParams`](https://github.com/mozilla/addons-frontend/blob/e680555435ba9cff8cd2144130e08bad526a6cea/src/core/utils/index.js#L170-L177) then you could delete this test. You would be delegating the responsibility of overriding params to that function and you could simply mock it to ensure the delegation.

File: tests/unit/amo/utils/test_index.js [hunk 3]
```diff
@@ -13,4 +15,211 @@ describe(__filename, () => {
...
+
+      expect(
+        _makeInternalExternalURL({
+          internalQueryParams,
+          urlString,
+        }),
+      ).toEqual(expectedURL);
+    });
+
+    it('overrides internal query params for an absolute URL for the host', () => {
```
Comment (kumar303) [inline tests/unit/amo/utils/test_index.js]: You could delete this test, too

File: tests/unit/amo/utils/test_index.js [hunk 4]
```diff
@@ -13,4 +15,211 @@ describe(__filename, () => {
...
+      expect(
+        _makeInternalExternalURL({
+          _config: getFakeConfig({ baseURL: siteBaseURL }),
+          externalQueryParams,
+          urlString,
+        }),
+      ).toEqual(expectedURL);
+    });
+
+    it('overrides external query params for an absolute URL for a different host', () => {
```
Comment (kumar303) [inline tests/unit/amo/utils/test_index.js]: ...and delete this one

File: src/amo/utils/index.js [hunk 5]
```diff
@@ -48,3 +50,34 @@ export const getCanonicalURL = ({
...
+  _config = config,
+  externalQueryParams = {},
+  internalQueryParams = {},
+  urlString,
+}: MakeInternalExternalURLParams) => {
+  const baseURL = _config.get('baseURL');
+  const urlParts = url.parse(urlString, true);
+
+  const queryParams =
+    !urlParts.protocol || baseURL.includes(urlParts.host)
```
Comment (kumar303) [inline src/amo/utils/index.js]: Did you consider just using this logic as a function like `isInternalUrl(url)`? I feel like that might make everything simpler (maybe the tests, too) but I guess it's fine as is. It was just a thought.

File: src/amo/utils/index.js [hunk 6]
```diff
@@ -48,3 +50,34 @@ export const getCanonicalURL = ({
...
+  _config: typeof config,
+  urlString: string,
+  internalQueryParams?: QueryParams,
+  externalQueryParams?: QueryParams,
+|};
+
+export const makeInternalExternalURL = ({
+  _config = config,
+  externalQueryParams = {},
+  internalQueryParams = {},
```
Comment (kumar303) [inline src/amo/utils/index.js]: maybe you should hard-code the default internal and external params and change the name of the function to something like `addParamsToPromoUrl()`?

**Comment by kumar303** [comment]
You could fix the bundlesize failure (https://travis-ci.org/mozilla/addons-frontend/jobs/578023555) by increasing the max value. Really, that test should only fail if the bundle increases by a *high percentage* (to catch unexpected inflation).

**Comment by bobsilverberg** [comment]
> r+wc
> 
> I'm still skeptical that we need a standalone function but you are deeper in the implementation of the two related features so I trust you 😄

Looking at your comments, and thinking about your initial comment, it's true that we could remove some tests and do some mocking if I made use of `addQueryParams`, so that does seem like a better option. If I do that then we probably don't need this function (as you originally suggested), so I'm going to probably just move it into the Secondary Hero component. Sorry for the noise.

**Comment by bobsilverberg** [comment]
I'm going to close this for now, implement the Secondary Hero without appending query params to the URLs, and then create a follow-up to that which will implement this same logic, but inside the Secondary Hero component.

**Pull request closed without merging**